### PR TITLE
Llama: device/type-invariant RoPE sin/cos computation, eager attention matches original implementation

### LINKED ISSUE #25681: LlamaRotaryEmbedding (wrong cache value when casting model to float16/bfloat16)
### System Info

- `transformers` version: 4.31.0
- Platform: Linux-5.15.0-79-generic-x86_64-with-glibc2.29
- Python version: 3.8.10
- Huggingface_hub version: 0.16.4
- Safetensors version: 0.3.2
- Accelerate version: 0.22.0.dev0
- Accelerate config:    - compute_environment: LOCAL_MACHINE
        - distributed_type: FSDP
        - mixed_precision: bf16
        - use_cpu: False
        - debug: False
        - num_processes: 1
        - machine_rank: 0
        - num_machines: 1
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - fsdp_config: {'fsdp_auto_wrap_policy': 'SIZE_BASED_WRAP', 'fsdp_backward_prefetch_policy': 'BACKWARD_PRE', 'fsdp_forward_prefetch': False, 'fsdp_min_num_params': 100000000, 'fsdp_offload_params': False, 'fsdp_sharding_strategy': 2, 'fsdp_state_dict_type': 'FULL_STATE_DICT', 'fsdp_sync_module_states': True, 'fsdp_use_orig_params': True}
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
        - dynamo_config: {'dynamo_backend': 'INDUCTOR'}
- PyTorch version (GPU?): 2.1.0.dev20230809+cu121 (True)
- Tensorflow version (GPU?): 2.13.0 (False)
- Flax version (CPU?/GPU?/TPU?): 0.7.2 (cpu)
- Jax version: 0.4.13
- JaxLib version: 0.4.13
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@ArthurZucker would be the best person to discuss this.

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

**TL;DR If a model with a `LlamaRotaryEmbedding` layer is cast to bfloat16/float16 after initialization and if during forward pass a sequence with a sequence length > `self.max_position_embeddings` is used, then the cached cos and sin buffer values will most probably be different than the trained model, giving unexpected results.**

I came across this very subtle error doing the following and I am not sure what might be the best solution for this.

I finetuned the Llama-2 model using accelerate FSDP and bfloat16 mixed precision policy. I used a slightly different config than the original one in which the `max_position_embeddings=2048` was set. FSDP + accelerate uses autocast under the hood which takes care of the ops inside `LlamaRotaryEmbedding` to be in full precision which is great. 

Problem happens when we feed a sequence with a greater sequence length and also cast the model to a lower precision as opposed to using autocast. I loaded this trained model using 

```python
load_checkpoint_and_dispatch(custom_config_model, str(fn),
                              device_map={
                                          "model":torch.cuda.current_device(),
                                          "lm_head":torch.cuda.current_device(),
                                         },
                              dtype=torch.bfloat16);
```

My custom config looked like this, notice `"max_position_embeddings": 2048,`:

```
 LlamaConfig {
   "block_size": 2960,
   "bos_token_id": 1,
   "eos_token_id": 2,
   "hidden_act": "silu",
   "hidden_size": 4096,
   "initializer_range": 0.02,
   "intermediate_size": 11008,
   "max_position_embeddings": 2048,
   "model_type": "llama",
   "num_attention_heads": 32,
   "num_hidden_layers": 32,
   "num_key_value_heads": 32,
   "packed_inputs": false,
   "pad_token_id": 0,
   "prefix_lm": false,
   "pretraining_tp": 1,
   "rms_norm_eps": 1e-06,
   "rope_scaling": null,
   "tie_word_embeddings": false,
   "transformers_version": "4.31.0",
   "use_cache": true,
   "vocab_size": 64008
 }
```

During inference when testing the trained model my training/validation perplexity increased from ~2.5 to ~20.0, it took me 2 days to figure out that the exact issue was with model casting + having sequence lengths > max_position_embeddings.

### Potential Fixes:

- Add warning about this, and suggest using autocast during inference.
- Add warning about this, and suggest initializing the model with a very high `self.max_position_embeddings` value so that cos-sin caches won't be re-initialized with wrong values due to lower precision. Even using, `self.max_position_embeddings=80k` should be fine given the relatively small size of the buffer compared to total model size.
- Modify `LlamaRotaryEmbedding` so that always float32 is used in ops and cast to `x.dtype` only at the very end. This is a bit difficult because if a model is cast to bfloat16/float16, it will still produce different cache values even if its cast back to float32. I don't know if there is way to disable model casting for certain layers - but I guess that would be autocast 😄 

This modified version will produce closer but still wrong cache values:

```python
class LlamaRotaryEmbedding(torch.nn.Module):
    def __init__(self, dim, max_position_embeddings=2048, base=10000, device=None):
        super().__init__()

        self.dim = dim
        self.max_position_embeddings = max_position_embeddings
        self.base = base
        inv_freq = 1.0 / (self.base ** (torch.arange(0, self.dim, 2).float().to(device) / self.dim))
        self.register_buffer("inv_freq", inv_freq, persistent=False)

        # Build here to make `torch.jit.trace` work.
        self._set_cos_sin_cache(
            seq_len=max_position_embeddings, device=self.inv_freq.device, dtype=torch.get_default_dtype()
        )

    def _set_cos_sin_cache(self, seq_len, device, dtype):
        self.max_seq_len_cached = seq_len
        t = torch.arange(self.max_seq_len_cached, device=device, dtype=dtype)

        freqs = torch.einsum("i,j->ij", t, self.inv_freq.to(dtype))
        # Different from paper, but it uses a different permutation in order to obtain the same calculation
        emb = torch.cat((freqs, freqs), dim=-1)
        self.register_buffer("cos_cached", emb.cos()[None, None, :, :].to(dtype), persistent=False)
        self.register_buffer("sin_cached", emb.sin()[None, None, :, :].to(dtype), persistent=False)

    def forward(self, x, seq_len=None):
        # x: [bs, num_attention_heads, seq_len, head_size]
        if seq_len > self.max_seq_len_cached:
            self._set_cos_sin_cache(seq_len=seq_len, device=x.device, dtype=torch.get_default_dtype())

        return (
            self.cos_cached[:, :, :seq_len, ...].to(dtype=x.dtype),
            self.sin_cached[:, :, :seq_len, ...].to(dtype=x.dtype),
        )
```

I personally will keep `self.max_position_embeddings` as high as my max intended sequence length and also will use autocast where possible.

### Reproduction

```python 
# from https://github.com/huggingface/transformers/blob/3d1edb6c5d36bf6426e72223f534266ff29c45c4/src/transformers/models/llama/modeling_llama.py#L92C1-L125C10

class LlamaRotaryEmbedding(torch.nn.Module):
    def __init__(self, dim, max_position_embeddings=2048, base=10000, device=None):
        super().__init__()

        self.dim = dim
        self.max_position_embeddings = max_position_embeddings
        self.base = base
        inv_freq = 1.0 / (self.base ** (torch.arange(0, self.dim, 2).float().to(device) / self.dim))
        self.register_buffer("inv_freq", inv_freq, persistent=False)

        # Build here to make `torch.jit.trace` work.
        self._set_cos_sin_cache(
            seq_len=max_position_embeddings, device=self.inv_freq.device, dtype=torch.get_default_dtype()
        )

    def _set_cos_sin_cache(self, seq_len, device, dtype):
        self.max_seq_len_cached = seq_len
        t = torch.arange(self.max_seq_len_cached, device=device, dtype=self.inv_freq.dtype)

        freqs = torch.einsum("i,j->ij", t, self.inv_freq)
        # Different from paper, but it uses a different permutation in order to obtain the same calculation
        emb = torch.cat((freqs, freqs), dim=-1)
        self.register_buffer("cos_cached", emb.cos()[None, None, :, :].to(dtype), persistent=False)
        self.register_buffer("sin_cached", emb.sin()[None, None, :, :].to(dtype), persistent=False)

    def forward(self, x, seq_len=None):
        # x: [bs, num_attention_heads, seq_len, head_size]
        if seq_len > self.max_seq_len_cached:
            self._set_cos_sin_cache(seq_len=seq_len, device=x.device, dtype=x.dtype)

        return (
            self.cos_cached[:, :, :seq_len, ...].to(dtype=x.dtype),
            self.sin_cached[:, :, :seq_len, ...].to(dtype=x.dtype),
        )
```

```python
# expected cache values
rotary_emb = LlamaRotaryEmbedding(2048)
rotary_emb.cos_cached[:,:,:1024]

tensor([[[[ 1.0000,  1.0000,  1.0000,  ...,  1.0000,  1.0000,  1.0000],
          [ 0.5403,  0.5478,  0.5552,  ...,  1.0000,  1.0000,  1.0000],
          [-0.4161, -0.3998, -0.3835,  ...,  1.0000,  1.0000,  1.0000],
          ...,
          [-0.9998,  0.9651, -0.8084,  ...,  0.9945,  0.9946,  0.9947],
          [-0.5550,  0.3096,  0.0407,  ...,  0.9945,  0.9946,  0.9947],
          [ 0.4001, -0.6259,  0.8536,  ...,  0.9945,  0.9946,  0.9947]]]])


# Wrong cache values when cast  to bfloat16
rotary_emb.to(torch.bfloat16);
# create an input > 2048
x = torch.randn(2, 32, 4096, 128)
_ = rotary_emb(x, seq_len=4096)
rotary_emb.cos_cached[:,:,:1024]

tensor([[[[ 1.0000,  1.0000,  1.0000,  ...,  1.0000,  1.0000,  1.0000],
          [ 0.5391,  0.5469,  0.5547,  ...,  1.0000,  1.0000,  1.0000],
          [-0.4160, -0.4023, -0.3809,  ...,  1.0000,  1.0000,  1.0000],
          ...,
          [-0.5273,  0.9180,  0.5625,  ...,  0.9961,  0.9961,  0.9961],
          [ 0.9883, -0.3008,  0.2578,  ...,  0.9961,  0.9961,  0.9961],
          [ 0.9883, -0.3008,  0.2578,  ...,  0.9961,  0.9961,  0.9961]]]])

# try with float16 this time
rotary_emb = LlamaRotaryEmbedding(2048)
# cast model to float16
rotary_emb.to(torch.float16);
rotary_emb.cos_cached[:,:,:1024]
# create an input > 2048
x = torch.randn(2, 32, 4096, 128)
_ = rotary_emb(x, seq_len=4096)
rotary_emb.cos_cached[:,:,:1024]

tensor([[[[ 1.0000,  1.0000,  1.0000,  ...,  1.0000,  1.0000,  1.0000],
          [ 0.5405,  0.5479,  0.5552,  ...,  1.0000,  1.0000,  1.0000],
          [-0.4163, -0.4001, -0.3831,  ...,  1.0000,  1.0000,  1.0000],
          ...,
          [-1.0000,  0.9185, -0.9453,  ...,  0.9946,  0.9946,  0.9946],
          [-0.5552,  0.1628, -0.2366,  ...,  0.9946,  0.9946,  0.9946],
          [ 0.4001, -0.7422,  0.6899,  ...,  0.9946,  0.9946,  0.9946]]]])
```

cc: @ArthurZucker 

### Expected behavior

Same cache values for rotary embeddings.

### LINKED ISSUE #27179: Llama inference instability in fp16 producing inf in the middle of the model
### System Info

- `transformers` version: 4.35.0.dev0
- Platform: Linux-5.15.0-1023-aws-x86_64-with-glibc2.31
- Python version: 3.9.16
- Huggingface_hub version: 0.17.3
- Safetensors version: 0.3.1
- Accelerate version: 0.25.0.dev0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.1.0+cu118 (True)
- Using GPU in script?: A100

### Who can help?

@ydshieh @fxmarty @gante

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Hi, I encounter inference instability with llama running in fp16 when left padding is used, and especially when full rows are masked out in the 4D attention mask.

At some point in the forward, `inf` values may appear in the intermediate logits, ultimately leading to tensors filled with `nan` and raising the error:

```
Traceback (most recent call last):
  File "=debug.py", line 38, in <module>
    outputs = model.generate(
  File "/fsx/felix/condaenvs/fx/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/fsx/felix/transformers/src/transformers/generation/utils.py", line 1704, in generate
    return self.sample(
  File "/fsx/felix/transformers/src/transformers/generation/utils.py", line 2822, in sample
    next_tokens = torch.multinomial(probs, num_samples=1).squeeze(1)
RuntimeError: probability tensor contains either `inf`, `nan` or element < 0
```

Note that the `inf` specifically appear at a padding position.

Reproduction:

```python
from transformers import AutoTokenizer, pipeline, logging, AutoModelForCausalLM
import torch

model_name_or_path = "meta-llama/Llama-2-7b-chat-hf"
token = "[specify your token]"

tokenizer = AutoTokenizer.from_pretrained(model_name_or_path, use_fast=True, token=token)
tokenizer.pad_token_id = tokenizer.eos_token_id
tokenizer.padding_side = "left"

with torch.device("cuda"):
    model = AutoModelForCausalLM.from_pretrained(model_name_or_path, torch_dtype=torch.float16, token=token)

sentence = "Felix Marty is a French"

# Alternatively, the issue can be reproduced with:
# sentence = "Elon Musk is a South"
# max_length=9

inp = tokenizer(sentence, return_tensors='pt', padding="max_length", max_length=9).to("cuda")

print("inp", inp["input_ids"].shape)
print("inp", inp)
torch.set_printoptions(threshold=10000000)

print("\n\n*** Generate:")
with torch.no_grad():
    outputs = model.generate(
        **inp,
        max_new_tokens=10,
        do_sample=True,
        top_p=0.9,
        temperature=float(0.01),
        top_k=40
    )

print(tokenizer.batch_decode(outputs))
```

Printing `torch.all(torch.isfinite())` at some points in the model, it appears the `inf` start to appear in the MLP at [`self.gate_proj(x)) * self.up_proj(x)`](https://github.com/huggingface/transformers/blob/3cd3eaf96048cb76e67a432e72a7cecbdd1630a8/src/transformers/models/llama/modeling_llama.py#L232) and things go crazy from there.

What's interesting is that for example fixing (two left padding tokens)
image

to

image

solves the issue.

It makes me think that the solution implemented for SDPA to avoid fully masked rows in the attention mask may actually be required for some other cases as this one https://github.com/huggingface/transformers/pull/26572 - but it is unclear why it relates to overflow here.

WDYT @gante @ydshieh? Is this something you have ever observed?

### Expected behavior

No `inf` spawning in the middle of inference with fp16 model

### LINKED ISSUE #28596: HfDeepSpeedConfig + ZeRO3 init accuracy bug!
### System Info

see https://github.com/microsoft/DeepSpeed/issues/4932

### Who can help?

_No response_

### Information

- [x] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

https://github.com/microsoft/DeepSpeed/issues/4932

### Expected behavior

https://github.com/microsoft/DeepSpeed/issues/4932

### LINKED ISSUE #28685: torch.arange use should not use dtype=float for integer ranges, conflicts w/ DS `zero.Init()`
### System Info

Impacts many versions of transformers up to and including current.

### Who can help?

@ArthurZucker @amyeroberts 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Use a number of transformers models that utilize arange for integer enumerations in the calculation of position embeddings with DeepSpeed zero.Init() and a low precision dtype (float16, bfloat16), and the generated embeddings will differ significantly from intended.

Using Llama as an example
`t = torch.arange(self.max_seq_len_cached, device=device, dtype=self.inv_freq.dtype)`

The inv_freq.dtype == float32. Single precision float can cover the required integer range for the enumeration (I believe it's in the 2k-8k range for Llama?).

However, when DeepSpeed zero.Init is used the init function patching will override the float dtype passed in with a low precision float dtype, so float32 -> bfloat16 or float16. Thus the integer range that can be represented without significant loss drops down to 256 for bfloat16 or 2048 for float16. DeepSpeed's patching has an exception for integer dtype, it will not cast arange to the low precision float dtype if arange dtype is an int type.

https://github.com/microsoft/DeepSpeed/blob/0dd0c615f8e6c7947ba81a4b0993284da5ec3209/deepspeed/runtime/zero/partition_parameters.py#L245-L246
```
def zero_wrapper_for_fp_tensor_constructor(fn: Callable, target_fp_dtype: torch.dtype) -> Callable:

    def wrapped_fn(*args, **kwargs) -> Tensor:
        if kwargs.get("device", None) is None:
            kwargs['device'] = torch.device(get_accelerator().device_name(os.environ["LOCAL_RANK"]))
        tensor: Tensor = fn(*args, **kwargs)
        if tensor.is_floating_point():
            tensor.data = tensor.data.to(target_fp_dtype)

        return tensor

    return wrapped_fn
```

torch.arange defaults to an integer dtype if start/end/step are ints. In this case though it's best to be explicit to make intent clear, we should explictly set dtype=torch.long (or torch.int64 depending on your tastes). Casting to float should be done after the arange. Additionally, in many position embedding calculation scenarios, it's best to try and keep the calculations in float32 as long as possible, doing final conversion to low precision type at the very end (if that's the dtype of inference or training).

### Expected behavior

Use of torch.arange should explicitly set dtype=torch.long (or int64).

Ex: for Llama, 

`t = torch.arange(self.max_seq_len_cached, device=device).type_as(self.inv_freq)`

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR fixes the following problems, all related to RoPE:
1. Casting a model with `.from_pretrained(..., torch_dtype=...)` or `.to(dtype=...)` would produce different sin/cos tensors at recomputation time. The underlying cause was `inv_freq` being a buffer, which means it was subject to buffer manipulation (like a `.to()` operation in the wrapping module). Note that [the original repo assumed it was always a `torch.float32` tensor](https://github.com/facebookresearch/llama/blob/ef351e9cd9496c579bf9f2bb036ef11bdc5ca3d2/llama/model.py#L100). In some models, there was a visible performance degradation when doing inference with `seq_len > max_position_embeddings` (see [here](https://github.com/huggingface/transformers/issues/28685#issuecomment-1921473526));
2. The `inv_freq` tensor was being loaded from the state dict, due to a previous version of the code where it was a persistent buffer;
3. ⚠️ Perhaps more importantly, the sin/cos tensors are now always computed on CPU. As pointed out in [this comment](https://github.com/microsoft/DeepSpeed/issues/4932#issuecomment-1911509172), there are subtle numerical differences that depend on the initialization device, which quickly escalate into further downstream issues. This particular change results in the following:
    a. Smaller modeling performance differences across devices, as CPUs are ubiquitous (as opposed to accelerators, which may change);
    b. Prevention of loss spikes at train time, possibly due to the more accurate sin/cos computation (see [this comment](https://github.com/microsoft/DeepSpeed/issues/4932#issuecomment-1902039818) and the whole issue);
    c. Slightly slower throughput when recomputing the sin/cos tensors, i.e. when going beyond `self.max_seq_len_cached`.
    
See additional data and experiments below for the impact of this PR. Most of the diff in this PR is tests, to ensure we don't regress 🤗 

Suggested review order:
1. Llama modelling changes
2. Llama test changes
3. GPTNeoX changes (fixes dtype cast as intended, see experiments below :) )
4. Other models (direct #Copied from changes)
5. Other tests (copy/paste)
(Other RoPE models will follow in a future PR)

## Related GH issues

Fixes https://github.com/huggingface/transformers/issues/28685
Fixes https://github.com/huggingface/transformers/issues/25681
Fixes https://github.com/huggingface/transformers/issues/28596
Fixes https://github.com/huggingface/transformers/issues/27179
Should fix/help https://github.com/microsoft/DeepSpeed/issues/4932

## Additional data and experiments

<details>
  <summary>Perlplexity, memory, and latency results before/after this PR</summary>

  NOTE: using the `.to()` casting method. The `torch_dtype` sees no differences, as `inv_freq` is not casted.

  <details>
    <summary>Llama 2 -- very little ppl differences</summary>

Dtype: `bfloat16`
(ignore the vram -- the latest commit has the same GPU memory footprint as `main`)

plot_perplexity_vram
plot_latency  

Dtype: `float16`
(ignore the vram -- the latest commit has the same GPU memory footprint as `main`)
plot_perplexity_vram
plot_latency

  </details>

  <details>
    <summary>TinyLlama -- visible ppl upgrade</summary>

Dtype: `bfloat16`
(ignore the vram -- the latest commit has the same GPU memory footprint as `main`)
plot_perplexity_vram
plot_latency

Dtype: `float16`
(ignore the vram -- the latest commit has the same GPU memory footprint as `main`)
plot_perplexity_vram
plot_latency

  </details>
  
</details>

<details>
  <summary>How sensible is the sin/cos creation to the device placement?</summary>
  
  Consider the following script:

  ```py
import torch
from transformers.models.llama.modeling_llama import LlamaRotaryEmbedding

TEST_DTYPE = torch.bfloat16

for dim in (64, 256, 1024):
    for max_position_embeddings in (1024, 2048, 4096):
        for base in (10000, 100000, 1000000):
            rope_gpu = LlamaRotaryEmbedding(dim=dim, max_position_embeddings=max_position_embeddings, base=base, device='cuda')
            rope_cpu = LlamaRotaryEmbedding(dim=dim, max_position_embeddings=max_position_embeddings, base=base, device='cpu')

            rope_cpu = rope_cpu.to(device='cuda', dtype=TEST_DTYPE)
            rope_gpu = rope_gpu.to(device='cuda', dtype=TEST_DTYPE)
            max_sin_diff = (rope_gpu.sin_cached - rope_cpu.sin_cached).abs().max()
            max_cos_diff = (rope_gpu.cos_cached - rope_cpu.cos_cached).abs().max()
            max_diff = max(max_sin_diff, max_cos_diff)
            if max_diff > 0.0:
                print(f"dim={dim}, max_position_embeddings={max_position_embeddings}, base={base}, max_diff={max_diff:.2e}")
```

On `main`, before this PR, we can see differences as large as ~`1e-3` regardless of `TEST_DTYPE` (even in `torch.float64`!). After this PR, the difference is `0.0`.
  
</details>

<details>
  <summary>Original Llama codebase vs our codebase after this PR?</summary>

Key takeaways:
👉 sin/cos are created on the available device (and not on CPU)
👉 sin/cos are not only kept in FP32, [but also applied in FP32](https://github.com/facebookresearch/llama/blob/ef351e9cd9496c579bf9f2bb036ef11bdc5ca3d2/llama/model.py#L156)!
  
Consider the following script, which compares this hugging face's implementation against [meta's repo](https://github.com/facebookresearch/llama)

  ```py
# run as `torchrun this_script.py`
from llama import Llama
from transformers import AutoModelForCausalLM
import torch

# Loaded in FP16 on GPU
original_llama = Llama.build(
    ckpt_dir="/home/joao/meta_llama/Llama-2-7b/",
    tokenizer_path="/home/joao/meta_llama/Llama-2-7b/tokenizer.model",
    max_seq_len=2048,  # internaly, 2048*2 is considered to compute sin/cos
    max_batch_size=1,
)
og_logits = original_llama.model(tokens=torch.tensor([list(range(1000))]), start_pos=0)
og_sin = original_llama.model.freqs_cis.imag
og_cos = original_llama.model.freqs_cis.real
del original_llama
torch.cuda.empty_cache()

# Loaded in FP16 on GPU
transformers_llama = AutoModelForCausalLM.from_pretrained(
    "meta-llama/Llama-2-7b-hf", device_map="auto", torch_dtype=torch.float16
)
logits = transformers_llama(torch.tensor([list(range(1000))])).logits.float()
sin = transformers_llama.model.layers[0].self_attn.rotary_emb.sin_cached
cos = transformers_llama.model.layers[0].self_attn.rotary_emb.cos_cached


logits_diff = (og_logits.cpu() - logits.cpu()).abs().max()
print(f"Max logits diff: {logits_diff.item()}")

# .cat -> our sin/cos have a period of 4pi (2 cycles), the orginal have a period of 2pi (1 cycle)
# .float() -> on main, we cast sin/cos to the model dtype
sin_diff = (torch.cat([og_sin, og_sin], dim=1).cpu() - sin.float().cpu()).abs().max()
cos_diff = (torch.cat([og_cos, og_cos], dim=1).cpu() - cos.float().cpu()).abs().max()
print(f"Max sin diff: {sin_diff.item()}")
print(f"Max cos diff: {cos_diff.item()}")
```

On `main` + GPU + FP16, before this PR, we can see sin/cos and logits differences as large as `2e-4` and `6e-2` (respectively). After this PR, the difference is `0.0`. 
  
</details>

### CODE REVIEW

File: tests/models/llama/test_modeling_llama.py [hunk 1]
```diff
@@ -505,6 +506,120 @@ def test_eager_matches_sdpa_generate(self):
             res_sdpa = model_sdpa.generate(**inputs, max_new_tokens=max_new_tokens, do_sample=False)
             self.assertTrue(torch.allclose(res_eager, res_sdpa))
 
+    @require_torch_gpu
+    def test_rope_cast_strategy_invariant(self):
```
Comment (gante) [inline tests/models/llama/test_modeling_llama.py:510]: this test fails on `main` because `inv_freq` was being casted with `.to()`
